File: sonar-server/src/main/webapp/WEB-INF/app/models/trends_chart.rb
#
# Sonar, entreprise quality control tool.
# Copyright (C) 2008-2012 SonarSource
# mailto:contact AT sonarsource DOT com
#
# Sonar is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3 of the License, or (at your option) any later version.
#
# Sonar is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with Sonar; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
#
class TrendsChart

  def self.png_chart(width, height, resource, metrics, locale, display_legend, options={})
    java_chart = Java::OrgSonarServerChartsJruby::TrendsChart.new(width, height, locale.to_s.gsub(/\-/, '_'), display_legend)

    init_series(java_chart, metrics)
    metric_ids=metrics.map{|m| m.id}
    add_measures(java_chart, time_machine_measures(resource, metric_ids, options))
    add_labels(java_chart, resource);

    export_chart_as_png(java_chart)
  end


  protected

  def self.init_series(java_chart, metrics)
    metrics.each do |metric|
      java_chart.initSerie(metric.id, metric.short_name, metric.val_type==Metric::VALUE_TYPE_PERCENT)
    end
  end

  def self.time_machine_measures(resource, metric_ids, options={})
    unless metric_ids.empty?
      sql= "select s.created_at as created_at, m.value as value, m.metric_id as metric_id, s.id as sid " +
            " from project_measures m LEFT OUTER JOIN snapshots s ON s.id=m.snapshot_id " +
            " where m.rule_id is null " +
            " and s.status=? " +
            " and s.project_id=? " +
            " and m.metric_id in (?) " +
            " and m.rule_priority is null and m.characteristic_id is null and m.person_id is null"
      if (options[:from])
        sql += ' and s.created_at>=?'
      end
      if (options[:to])
        sql += ' and s.created_at<=?'
      end
      sql += ' order by s.created_at ASC'
      conditions=[sql, Snapshot::STATUS_PROCESSED, resource.id, metric_ids]
      if (options[:from])
        conditions<<options[:from]
      end
      if (options[:to])
        conditions<<options[:to]
      end
      ProjectMeasure.connection.select_all(Project.send(:sanitize_sql, conditions))
    end
  end
   
  def self.add_measures(java_chart, sqlresult)
    if sqlresult && sqlresult.size>0
      sqlresult.each do |hash|
        value = hash["value"].to_f
        time = Time.parse( hash["created_at"].to_s )
        serie_id=hash["metric_id"].to_i
        java_chart.addMeasure(value, time, serie_id)
      end
    end
  end
  
  def self.add_labels(java_chart, resource)
    categories=EventCategory.categories(true)
    category_names=categories.map{|cat| cat.name}
    Event.find(:all, :conditions => {:resource_id => resource.id}).each do |event|
      if category_names.include?(event.category)
        time = Time.parse( event.event_date.to_s )
        java_chart.addLabel(time, event.name, event.category==EventCategory::KEY_ALERT)
      end
    end
  end
  
  def self.export_chart_as_png(java_chart)
    java_int_array_to_bytes(java_chart.exportChartAsPNG())
  end

  def self.java_int_array_to_bytes( java_int_array )
    # pack does not exists on the return java array proxy object
    # the java proxy array contains integers, they must be packed
    # to return a correct byte array stream (c*)
    [].concat(java_int_array).pack("c*")
  end

end

File: sonar-server/src/main/java/org/sonar/server/charts/jruby/TrendsChart.java
/*
 * Sonar, open source software quality management tool.
 * Copyright (C) 2009 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * Sonar is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Sonar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Sonar; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package org.sonar.server.charts.jruby;

import org.apache.commons.lang.LocaleUtils;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.axis.AxisLocation;
import org.jfree.chart.axis.DateAxis;
import org.jfree.chart.axis.NumberAxis;
import org.jfree.chart.plot.Marker;
import org.jfree.chart.plot.ValueMarker;
import org.jfree.chart.plot.XYPlot;
import org.jfree.chart.renderer.xy.XYLineAndShapeRenderer;
import org.jfree.chart.title.TextTitle;
import org.jfree.data.time.Day;
import org.jfree.data.time.TimeSeries;
import org.jfree.data.time.TimeSeriesCollection;
import org.jfree.ui.RectangleAnchor;
import org.jfree.ui.RectangleEdge;
import org.jfree.ui.RectangleInsets;
import org.jfree.ui.TextAnchor;
import org.sonar.server.charts.deprecated.BaseChart;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.text.DateFormat;
import java.text.DecimalFormat;
import java.text.ParseException;
import java.util.Date;
import java.util.TreeMap;

public class TrendsChart extends BaseChart {

  private XYPlot plot;
  private TreeMap<Long, TimeSeries> seriesById;
  private int percentAxisId = -1;
  private boolean displayLegend;

  public TrendsChart(int width, int height, String localeKey, boolean displayLegend) {
    super(width, height);
    this.displayLegend = displayLegend;
    seriesById = new TreeMap<Long, TimeSeries>();
    plot = new XYPlot();
    DateAxis dateAxis = new DateAxis();
    dateAxis.setDateFormatOverride(DateFormat.getDateInstance(DateFormat.SHORT, LocaleUtils.toLocale(localeKey)));
    plot.setDomainAxis(dateAxis);
  }

  public void initSerie(Long serieId, String legend, boolean isPercent) {
    TimeSeries series = new TimeSeries(legend);

    int index=seriesById.size();
    seriesById.put(serieId, series);

    TimeSeriesCollection timeSeriesColl = new TimeSeriesCollection();
    timeSeriesColl.addSeries(series);
    plot.setDataset(index, timeSeriesColl);

    if (isPercent) {
      if (percentAxisId == -1) {
        NumberAxis rangeAxis = new NumberAxis();
        rangeAxis.setNumberFormatOverride(new DecimalFormat("0'%'"));
        rangeAxis.setStandardTickUnits(NumberAxis.createIntegerTickUnits());
        rangeAxis.setUpperBound(100.0);
        rangeAxis.setLowerBound(0.0);
        plot.setRangeAxisLocation(index, AxisLocation.TOP_OR_LEFT);
        plot.setRangeAxis(index, rangeAxis);
        plot.mapDatasetToRangeAxis(index, index);
        percentAxisId = index;

      } else {
        plot.mapDatasetToRangeAxis(index, percentAxisId);
      }
    } else {
      NumberAxis rangeAxis = new NumberAxis(displayLegend ? legend : null);
      rangeAxis.setAutoRangeIncludesZero(false);
      rangeAxis.setStandardTickUnits(NumberAxis.createIntegerTickUnits());
      rangeAxis.setAutoRangeMinimumSize(2.0);
      plot.setRangeAxisLocation(index, AxisLocation.TOP_OR_RIGHT);
      plot.setRangeAxis(index, rangeAxis);
      plot.mapDatasetToRangeAxis(index, index);
    }

    XYLineAndShapeRenderer renderer = new XYLineAndShapeRenderer();
    renderer.setBaseShapesVisible(false);
    renderer.setSeriesStroke(0, new BasicStroke(2.0f));
    renderer.setSeriesPaint(0, COLORS[index % COLORS.length]);
    plot.setRenderer(index, renderer);
  }

  public void addMeasure(Double value, Date date, Long serieId) {
    seriesById.get(serieId).addOrUpdate(new Day(date), value);
  }

  public void addLabel(Date date, String label) throws ParseException {
    addLabel(date, label, false);
  }

  public void addLabel(Date date, String label, boolean lower) throws ParseException {
    Day d = new Day(date);
    double millis = d.getFirstMillisecond();
    Marker marker = new ValueMarker(millis);
    marker.setLabel(label);
    marker.setLabelAnchor(RectangleAnchor.TOP_RIGHT);
    marker.setLabelTextAnchor(TextAnchor.TOP_LEFT);
    Color c = new Color(17, 40, 95);
    marker.setLabelPaint(c);
    marker.setPaint(c);
    marker.setStroke(new BasicStroke(1.0f, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND, 3.0f, new float[]{5f, 5f, 5f, 5f}, 2.0f));
    if (lower) {
      marker.setLabelOffset(new RectangleInsets(18, 0, 0, 5));
    }
    plot.addDomainMarker(marker);
  }

  @Override
  protected BufferedImage getChartImage() throws IOException {
    JFreeChart chart = new JFreeChart(null, TextTitle.DEFAULT_FONT, plot, true);
    configureChart(chart, displayLegend ? RectangleEdge.BOTTOM : null);
    return super.getBufferedImage(chart);
  }
}